docs/LexerDevelopment.md

```diff
@@ -0,0 +1,377 @@
+   expression matches.
+
+In the above example, there are two rules. The first rule yields the token
+`Text::Whitespace` but does not do anything to the state stack. The second rule
+yields the token `Str::Double` and adds the new state `:string` to the top of
+the state stack. The text being parsed after this point will now be processed by
+the rules in the `:string` state.
+
+The following code shows the definition for this state and the rules that are
+defined within it:
+
+```rb
+state :string do
+  rule /[^\\"]+/, Str::Double
+  rule /\\./, Str::Escape
+  rule /"/, Str::Double, :pop!
+end
+```
+
+The last rule features the "special state" `:pop!`. This is not really a state,
+rather it is an instruction to the lexer to remove the current state from the
+top of the state stack. In the JSON lexer, when we encounter the double
+quotation mark `"` we enter into the state of being "in a string" and when we
+next encounter the double quotation mark, we leave the string and return to the
+previous state (in this case, the `:root` state).
+
+*Complex Rules*
+
+It is possible to define more complex rules for a lexer by calling `rule` with:
+
+1. a regular expression to match against; and
+2. a block to call if the regular expression matches.
+
+The block called can take one argument, usually written as `m`, that contains
+the regular expression match.
+
+These kind of rules allow for more fine-grained control of the state stack.
+Inside a complex rule's block, it's possible to {Rouge::RegexLexer#push push},
+{Rouge::RegexLexer#pop! pop}, {Rouge::RegexLexer#token yield a token} and 
+{Rouge::RegexLexer#delegate delegate to another lexer}.
+
+You can see an example of these more complex rules in [the Ruby
+lexer][ruby-lexer].
+
+### Additional Features
+
+While the properties and states are the minimum elements of a lexer that should
+be implemented, a lexer can include additional features.
+
+#### Source Detection 
+
+```rb
+def self.detect?(text)
+  return true if text.shebang? 'ruby'
+end
+```
+Rouge will attempt to guess the appropriate lexer if it is not otherwise clear.
+If Rouge is unable to do this on the basis of any tag, associated filename or
+associated mimetype, it will try to detect the appopriate lexer on the basis of
+the text itself (the source). This is done by calling `self.detect?` on the
+possible lexer (a default `self.detect?` method is defined in {Rouge::Lexer
+Lexer} and simply returns `false`).
+
+A lexer can implement its own `self.detect?` method that takes as a parameter a
+{Rouge::TextAnalyzer TextAnalyzer} object. If the `self.detect?` method
+returns true, the lexer will be selected as the appropriate lexer.
+
+The `self.detect?` method is intended to work by looking at the shebang or
+doctype that identifies a piece of text. To make this easier, Rouge provides
+the {Rouge::TextAnalyzer#shebang TextAnalyzer#shebang} method and the
+{Rouge::TextAnalyzer#doctype TextAnalyzer#doctype} method.
```

Comment:
@pyrmont What does the following mean in the lines above and why are they within braces? YARD syntax?
- `{Rouge::TextAnalyzer#shebang TextAnalyzer#shebang}`
- `{Rouge::TextAnalyzer#doctype TextAnalyzer#doctype}`
(They look weird via GitHub UI)

Reply:
@ashmaroli Yep, those are [YARD links](https://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md#Linking_Objects______). It's sort of an added bonus that because YARD can process Markdown files, documentation files can be read (relatively) easily on GitHub itself.

I'll make a slight tweak to avoid the alternative title for the links, though. When reading on RubyDoc.info, I prefer to omit the top-most module name (so that you see Lexer rather than Rouge::Lexer) but when looked at on GitHub, it does make it quite difficult to parse.